ifplugd-0.28/man/ifplugstatus.8
.TH ifplugstatus 8 User Manuals
.SH NAME
ifplugstatus \- A link beat detection tool
.SH SYNOPSIS
\fBifplugstatus [\fIoptions\fB] [\fIINTERFACE\fB]
\f1
.SH DESCRIPTION
ifplugstatus is an utility which may be used to detect the link status of a local Linux ethernet device, much in the same way mii-diag, mii-tool and ethtool work. In fact it supports all three different APIs these three tools use. In addition it supports link checking with the IFF_RUNNING interface flag, which most modern drivers (not only ethernet) support, and association status checking with the wireless extension API for WLAN devices.

The APIs are tried in the following order:

First the newest API: \fISIOCETHTOOL\f1 (aka ethtool API)

Second the next older API: \fISIOCGMIIREG\f1 (aka mii-diag API)

Than the WLAN API: \fIwireless extension\f1 (aka iwconfig API)

Followed by the cleanest API: \fIIFF_RUNNING\f1 (aka ifconfig API)

The oldest API (\fISIOCPRIV\f1 aka mii-tool API) is not tested since it is obsolete.

ifplugstatus may be used in shell script since it returns the current status as return value. It is especially useful to detect the available APIs on the used network driver. (Option \fB-v\f1)
.SH OPTIONS
You may specify an ethernet device on the command line. Otherwise ifplugstatus will check all available network interfaces.
.TP
\fB-a | --auto\f1
Enable interface automatically before querying (default: off) 
.TP
\fB-h | --help\f1
Show help 
.TP
\fB-q | --quiet\f1
Decrease verbosity by one. If the verbosity is < 0, no text will be shown, only the return value is relevant; if the verbosity is = 0, a terse status will be shown; If the verbosity is > 0, detailed information about the used API is returned. (By default the verbosity is 0)
.TP
\fB-v | --verbose\f1
Increase verbosity by one. See option \fB-q\f1. 
.TP
\fB-V | --version\f1
Show version 
.SH RETURN VALUES
\fI0\f1 Success

\fI1\f1 Failure

\fI2\f1 Link beat detected (only available when an interface is specified)

\fI3\f1 Unplugged (same here)
.SH AUTHOR
ifplugd was written by Lennart Poettering <mzvscyhtq (at) 0pointer (dot) de>. ifplugd is available at \fBhttp://0pointer.de/lennart/projects/ifplugd/\f1
.SH SEE ALSO
\fBmii-diag(8)\f1, \fBmii-tool(8)\f1, \fBethtool(8)\f1, \fBifplugd(8)\f1
.SH COMMENTS
This man page was written using \fBxml2man(1)\f1 by Oliver Kurth.


ifplugd-0.28/src/ifplugstatus.c
/* $Id: ifplugstatus.c 114 2004-12-19 00:08:01Z lennart $ */

/*
 * This file is part of ifplugd.
 *
 * ifplugd is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * ifplugd is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with ifplugd; if not, write to the Free Software Foundation,
 * Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <stdio.h>
#include <errno.h>
#include <string.h>
#include <unistd.h>
#include <getopt.h>
#include <stdlib.h>
#include <sys/ioctl.h>
#include <net/if.h>

#include <libdaemon/dlog.h>

#include "interface.h"
#include "svn-revision.h"

int interface_auto_up = 0, interface_do_message = 0;

int verbose = 0;
char *interface = NULL;

int handle(char *iface) {
    int fd, r = 0;
    interface_status_t s;
    
    if ((fd = socket(PF_INET, SOCK_DGRAM, 0)) < 0)
        return -1;
    
    if (verbose > 0) {
        printf("%s:\n", iface);
        
	if ((s = interface_detect_beat_ethtool(fd, iface)) == IFSTATUS_ERR)
            printf("    SIOCETHTOOL failed (%s)\n", strerror(errno));
        else
            printf("    SIOCETHTOOL: %s\n", s == IFSTATUS_UP ? "link beat detected" : "unplugged");
        
        if ((s = interface_detect_beat_mii(fd, iface)) == IFSTATUS_ERR)
            printf("    SIOCGMIIPHY failed (%s)\n", strerror(errno));
        else
            printf("    SIOCGMIIPHY: %s\n", s == IFSTATUS_UP ? "link beat detected" : "unplugged");

        if ((s = interface_detect_beat_wlan(fd, iface)) == IFSTATUS_ERR)
            printf("    Wireless failed.\n");
        else
            printf("    Wireless: %s\n", s == IFSTATUS_UP ? "link beat detected" : "unplugged");

        if ((s = interface_detect_beat_iff(fd, iface)) == IFSTATUS_ERR)
            printf("    IFF_RUNNING failed.\n");
        else
            printf("    IFF_RUNNING: %s\n", s == IFSTATUS_UP ? "link beat detected" : "unplugged");
        
        if ((s = interface_detect_beat_priv(fd, iface)) == IFSTATUS_ERR)
            printf("    SIOCDEVPRIVATE failed (%s)\n", strerror(errno));
        else
            printf("    SIOCDEVPRIVATE: %s\n", s == IFSTATUS_UP ? "link beat detected" : "unplugged");

    } else {
        
        if ((s = interface_detect_beat_ethtool(fd, iface)) == IFSTATUS_ERR)
            if ((s = interface_detect_beat_mii(fd, iface)) == IFSTATUS_ERR)
                if ((s = interface_detect_beat_wlan(fd, iface)) == IFSTATUS_ERR)
                    s = interface_detect_beat_iff(fd, iface);

        switch(s) {
            case IFSTATUS_UP:
                if (!verbose)
                    printf("%s: link beat detected\n", iface);
                r = 1;
                break;
                
            case IFSTATUS_DOWN:
                if (!verbose)
                    printf("%s: unplugged\n", iface);
                r = 2;
                break;

            default:
                if (!verbose)
                    printf("%s: not supported%s\n", iface, getuid() != 0 ? " (Retry as root?)" : "");
                
                r = -1;
                break;
        }
    }
    
            
    close(fd);
    return r;
}

void usage(char *p) {
    if (strrchr(p, '/'))
        p = strchr(p, '/')+1;

    printf("%s [options] [INTERFACE]\n"
           "   -a --auto            Enable interface automatically (%s)\n"
           "   -q --quiet           Quiet behaviour (%s)\n"
           "   -v --verbose         Enable verbosity (%s)\n"
           "   -h --help            Show this help\n"
           "   -V --version         Show version number\n",
           p, interface_auto_up ? "on" : "off", verbose < 0 ? "on" : "off", verbose > 0 ? "on" : "off");
}


void parse(int argc, char *argv[]) {
    static struct option long_options[] = {
        {"auto",        no_argument, 0, 'a'},
        {"quiet",       no_argument, 0, 'q'},
        {"verbose",     no_argument, 0, 'v'},
        {"help",        no_argument, 0, 'h'},
        {"version",     no_argument, 0, 'V'},
        {0, 0, 0, 0}
    };
    int option_index = 0;
    int help = 0;
    
    for (;;) {
        int c; 
       
        if ((c = getopt_long(argc, argv, "avhqV", long_options, &option_index)) < 0)
            break;

        switch (c) {
            case 'a' :
                interface_auto_up = !interface_auto_up;
                break;
            case 'v':
                verbose++; 
                break;
            case 'q':
                verbose--; 
                break;
            case 'h':
                help = 1;
                break;
            case 'V':
#ifdef SVN_REVISION
		 printf("ifplugstatus "VERSION" (SVN: "SVN_REVISION")\n");
#else
		 printf("ifplugstatus "VERSION"\n");
#endif
                exit(0);
            default:
                fprintf(stderr, "Unknown parameter.\n");
                exit(1);

        }
    }

    if (help) {
        usage(argv[0]);
        exit(0);
    }

    if (optind < argc)
        interface = argv[optind];
}


int main(int argc, char *argv[]) {
    parse(argc, argv);

    if (interface) {
        int r;

        if ((r = handle(interface)) < 0) {
            if (verbose == 0)
                fprintf(stderr, "Failure (%s)\n", strerror(errno));
            return 1;
        }

        return r+1;
        
    } else {
        FILE *f;
        char ln[256];

        if (!(f = fopen("/proc/net/dev", "r"))) {
            fprintf(stderr, "Failed to open /proc/net/dev: %s\n", strerror(errno));
            return 1;
        }

        fgets(ln, sizeof(ln), f);
        fgets(ln, sizeof(ln), f);

        while (fgets(ln, sizeof(ln), f)) {
            char *p, *e;

            p = ln+strspn(ln, " \t");
            if (!(e = strchr(p, ':'))) {
                fprintf(stderr, "Parse failure in /proc/net/dev.\n");
                fclose(f);
                return 1;
            }

            *e = 0;
            handle(p);
        }
        
        fclose(f);
    }
    
    return 0;
}
